src/build.rs

```diff
@@ -321,92 +321,80 @@ impl BuildQueue {
+
+                if self.config.cfg_test {
+                    // FIXME(#351) allow passing --test to lib crate-type when building a dependency
+                    if is_final_crate_type {
                         args.push("--test".to_owned());
+                    } else {
+                        args.push("--cfg".to_owned());
+                        args.push("test".to_owned());
                     }
-                    if self.config.sysroot.is_empty() {
-                        args.push("--sysroot".to_owned());
-                        let home = env::var("RUSTUP_HOME").or(env::var("MULTIRUST_HOME"));
-                        let toolchain = env::var("RUSTUP_TOOLCHAIN").or(env::var("MULTIRUST_TOOLCHAIN"));
-                        let sys_root = if let (Ok(home), Ok(toolchain)) = (home, toolchain) {
-                            format!("{}/toolchains/{}", home, toolchain)
-                        } else {
-                            env::var("SYSROOT")
-                                .map(|s| s.to_owned())
-                                    .ok()
-                                    .or_else(|| Command::new(&rustc_exe)
-                                        .arg("--print")
-                                        .arg("sysroot")
-                                        .output()
-                                        .ok()
-                                        .and_then(|out| String::from_utf8(out.stdout).ok())
-                                        .map(|s| s.trim().to_owned()))
-                                    .expect("need to specify SYSROOT env var, \
-                                            or use rustup or multirust")
-                        };
-                        args.push(sys_root.to_owned());
+                }
+                if self.config.sysroot.is_empty() {
+                    let sysroot = current_sysroot()
+                                    .expect("need to specify SYSROOT env var or use rustup or multirust");
+                    args.push("--sysroot".to_owned());
+                    args.push(sysroot);
+                }
+
+                // We can't omit compilation here, because Cargo is going to expect to get
+                // dep-info for this crate, so we shell out to rustc to get that.
+                // This is not really ideal, because we are going to
+                // compute this info anyway when we run rustc ourselves, but we don't do
+                // that before we return to Cargo.
+                // FIXME Don't do this. Start our build here rather than on another thread
+                // so the dep-info is ready by the time we return from this callback.
+                // Since ProcessBuilder doesn't allow to modify args, we need to create
+                // our own command here from scratch here.
+                let rustc_exe = env::var("RUSTC").unwrap_or("rustc".to_owned());
```

Comment:
It makes me a bit sad to duplicate this code - given the string literals it feels like an easy error for someone to modify one instance without modifying the other.